core/types/rollup_cost.go

```diff
@@ -114,6 +123,42 @@ func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
 	}
 	forBlock := ^uint64(0)
 	var cachedFunc l1CostFunc
+	selectFunc := func(blockTime uint64) l1CostFunc {
+		// Note: the various state variables below are not initialized from the DB until this
+		// point to allow deposit transactions from the block to be processed first by state
+		// transition.  This behavior is consensus critical!
+		l1FeeScalars := statedb.GetState(L1BlockAddr, L1FeeScalarsSlot).Bytes()
+		l1BlobBaseFee := statedb.GetState(L1BlockAddr, L1BlobBaseFeeSlot).Big()
+		l1BaseFee := statedb.GetState(L1BlockAddr, L1BaseFeeSlot).Big()
+
+		if config.IsOptimismEcotone(blockTime) {
+			// Edge case: the very first Ecotone block requires we use the Bedrock cost
+			// function. We detect this scenario by checking if the Ecotone parameters are
+			// unset. Note here we rely on assumption that the scalar parameters are adjacent
+			// in the buffer and l1BaseFeeScalar comes first. We need to check this prior to
+			// other forks, as the first block of Fjord and Ecotone could be the  same block.
+			firstEcotoneBlock := l1BlobBaseFee.BitLen() == 0 &&
+				bytes.Equal(emptyScalars, l1FeeScalars[scalarSectionStart:scalarSectionStart+8])
```

Comment:
Wondering why we don't do something similar like for op-node and add a `IsOptimismFirstEcotoneBlock(uint64) bool` to the config. In the monorepo we have `IsEcotoneActivationBlock`, which is more straight forward and doesn't check indirect assumptions.

Reply:
I believe the reason we checked these assumptions when we did the Ecotone upgrade vs. having a similar function, is that we don't have access to the block time param in the op-geth genesis config. (op-nodes does via the rollup config).

We could add it to the genesis.json file (under config, along w/ block fork times), but it would require modification for all chains running Ecotone.